src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -18,24 +20,44 @@ namespace Microsoft.AspNetCore.Routing
     [DebuggerDisplay("{DebuggerDisplayString,nq}")]
     public sealed class CompositeEndpointDataSource : EndpointDataSource
     {
-        private readonly EndpointDataSource[] _dataSources;
         private readonly object _lock;
+        private readonly ObservableCollection<EndpointDataSource> _dataSources;
         private IReadOnlyList<Endpoint> _endpoints;
         private IChangeToken _consumerChangeToken;
         private CancellationTokenSource _cts;
 
-        internal CompositeEndpointDataSource(IEnumerable<EndpointDataSource> dataSources)
+        internal CompositeEndpointDataSource()
         {
-            if (dataSources == null)
-            {
-                throw new ArgumentNullException(nameof(dataSources));
-            }
+            _dataSources = new ObservableCollection<EndpointDataSource>();
+            _dataSources.CollectionChanged += OnDataSourcesChanged;
 
             CreateChangeToken();
-            _dataSources = dataSources.ToArray();
             _lock = new object();
         }
 
+        // TODO: Used by old tests. Remove in future
+        internal CompositeEndpointDataSource(IEnumerable<EndpointDataSource> endpointDataSources) : this()
+        {
+            foreach (var dataSource in endpointDataSources)
+            {
+                _dataSources.Add(dataSource);
+            }
+        }
+
+        private void OnDataSourcesChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
+            lock (_lock)
+            {
+                // Only trigger changes if composite data source has already initialized endpoints
+                if (_endpoints != null)
+                {
+                    HandleChange();
+                }
+            }
+        }
+
+        public IList<EndpointDataSource> DataSources => _dataSources;
```

Comment:
Consider instead making the composite Implement `ICollection` - if this ends up ever going public, that just feels a little more first class.